spec/src/main/asciidoc/architecture.asciidoc

```diff
@@ -193,47 +193,442 @@ For example, calling an `@Outgoing` annotated method from user code will not pos
+
+|
+[source, java]
+----
+@Outgoing("channel") 
+CompletionStage<O> method()
+----
+| Produces a infinite streams of _payload_ associated with the channel `channel`. Produced payloads are mapped to `Message<O>` by the Reactive Messaging implementation. The result is a `CompletionStage`. The method cannot be called until the `CompletionStage` returned previously is completed.
+|	This method is called for each _request_ made by the subscriber.
+
+|===
+
+
+===== Methods consuming data
+
+[cols="2a,1,1",options="header"]
+|===
+|Signature
+|Behavior
+|Invocation
+
+|
+[source,java]
+----
+@Incoming("channel")
+Subscriber<Message<I>> method()
+----
+| Returns a `Subscriber` that receives the `Message` objects transiting on the channel `channel`.
+| The method is called only once to retrieve the `Subscriber` object at assembly time.
+ This subscriber is connected to the matching channel.
+
+|
+[source,java]
+----
+@Incoming("channel")
+Subscriber<I> method()
+----
+| Returns a `Subscriber` that receives the _payload_ objects transiting on the channel `channel`.
+The payload is automatically extracted from the inflight messages using `Message.getPayload()`.
+| The method is called only once to retrieve the `Subscriber` object at assembly time.
+ This subscriber is connected to the matching channel.
+
+|
+[source,java]
+----
+@Incoming("channel") 
+SubscriberBuilder<Message<I>> method()
+----
+| Returns a `SubscriberBuilder` that receives the `Message` objects transiting on the channel `channel`.
+| The method is called only once to retrieve the `Subscriber` object at assembly time. 
+The created `subscriber` is connected to the matching channel.
+
+|
+[source,java]
+----
+@Incoming("channel")
+SubscriberBuilder<I> method()
+----
+| Returns a `SubscriberBuilder` that receives the `Message` objects transiting on the channel `channel`.
+The payload is automatically extracted from the inflight messages using `Message.getPayload()`.
+| The method is called only once to retrieve the `Subscriber` object at assembly time.
+The created `subscriber` is connected to the matching channel.
+
+|
+[source,java]
+----
+@Incoming("channel")
+void method(I payload)
+----
+| Consumes the _payload_.
+The method can return `void` or any object or `null`. 
+The returned value is ignored. 
+| This method is called for every `Message<I>` instances transiting on the channel `channel`. 
+The payload is automatically extracted from the inflight messages using `Message.getPayload()`.
+The user method is never called concurrently and so must return before being called with the next payload.
+
+|
+[source,java]
+----
+@Incoming("channel")
+CompletionStage<?> method(Message<I> msg)
+----
+| Consumes the `Message` 
+| This method is called for every `Message<I>` instances transiting on the channel `channel`. 
+The user method is never called concurrently and so the returned `CompletionStage` must be completed before being calling the method with the next `Message`.
```

Comment:
This is not actually a very logically clear statement IMO. The timeframe of the method invocation obviously can end prior to the CompletionStage being resolved. This is an `@Incoming` so we are providing the role of a Subscriber. We do not need the result from the CompletionStage in the same way that we would want the result from a Processor. Do we? Is there anyway that we want to be able to access the Streams CompletionSubscriber/CompletionStage in some middleware value add? If the user wishes to use a CompletionStage should we not allow multiple as yet unresolved Subscriber onData results?  Also, what about the result of the _completed_ stream - onComplete() is that just unavailable?

Reply:
The `CompletionStage` is there to indicate when the incoming message has been processed correctly.  As the message can be asynchronously processed, the returned `CompletionStage` is used to know when the task has completed (or failed). 

But yes, the value used to complete the returned CompletionStage is meaningless (and discarded), and there is no way (at the moment) to retrieve the value. 

The completion of the `CompletionStage` controls the flow as we won't be calling this method before having received the completion of the previous call.

Reply:
I've updated the content to explain more explicitly what's the expected behavior.